### PR TITLE
creator as non-classic data object

### PR DESCRIPTION
**Pull request by abaevbog**
- Removed ClassicDataObject as a parent class of Zotero_Creators
- Removed all `load()`-ing of Zotero_Creators. An object is constructed with data right away. 
- One `join` in `loadCreators` to load all creators on GET requests, instead of sequential `$item->load()`.
- Some minor refactoring to try to avoid extra queries

### CODE REVIEW

**Comment by abaevbog** [comment]
One of the consequences of removing `libraryID` from the `creators` table is that the items from the same shard essentially "share" creators across libraries. So if 2 items are created with the same creator's information, we have 1 `creator` entry, and 2 `itemCreators` entries linking this creator to items.

I suppose this is a good thing, since we don't have to store duplicate data about authors across libraries. But something to account for is how to deal with the deletion of obsolete creators. On every delete of an `item`, we can check if its creator belongs to any other `items`, and delete them if not. Or, we can run a cron job every 30 minutes or so to fetch and delete `creators` that are not linked to any items.

I think the second option is better, since the first option requires performing extra queries on every single DELETE Api call, which puts extra load on the dataserver

**Comment by dstillman** [comment]
Sorry, should've noted this earlier — we should just do what we do for `itemData` and inline the values in `itemCreators`. That'll mean a bit of data duplication, but it'll be faster, and we really don't want to have to purge these in an event.

(The empty `itemDataValueHash` column in `itemData` is a holdover from many years ago when we had a separate `itemDataValues` table like the client.)

File: misc/db-updates/2023-07-17/classic_data_objects.sql [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+DROP trigger if exists fku_itemCreators_libraryID;
+
+CREATE TABLE `creators` (
+  `creatorID` bigint unsigned NOT NULL AUTO_INCREMENT,
+  `firstName` varchar(255) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci DEFAULT NULL,
+  `lastName` varchar(255) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci DEFAULT NULL,
+  `fieldMode` tinyint(1) unsigned DEFAULT NULL,
+  PRIMARY KEY (`creatorID`),
+  KEY `name` (`lastName`(7),`firstName`(6))
+) ENGINE=InnoDB DEFAULT CHARSET=utf8;
```
Comment (dstillman) [inline misc/db-updates/2023-07-17/classic_data_objects.sql:14]: This will need to be a PHP script like the others that does a proper migration of existing data. It should set the shard to `down`, wait 20 seconds, do the migration, and bring the shard up.

File: misc/master.sql [hunk 2]
```diff
@@ -184,6 +184,7 @@ CREATE TABLE `libraries` (
   `lastUpdated` timestamp NOT NULL DEFAULT '0000-00-00 00:00:00',
   `version` int(10) unsigned NOT NULL DEFAULT '0',
   `shardID` smallint(5) unsigned NOT NULL,
+  `hasData` TINYINT( 1 ) NOT NULL DEFAULT '0',
```
Comment (dstillman) [inline misc/master.sql:187]: We need to add this, but it shouldn't be in this PR.

Comment (abaevbog) [reply misc/master.sql:187]: Can I push just this one line to master?

Comment (dstillman) [reply misc/master.sql:187]: Sure.

**Comment by abaevbog** [comment]
> inline the values in `itemCreators`

Got it. So then, we completely drop the `creators` table, and the `itemsCreators` will become something like this:

CREATE TABLE itemCreators (
  'itemID' bigint unsigned NOT NULL,
  'firstName' varchar(255) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  'lastName' varchar(255) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  'fieldMode' tinyint(1) unsigned DEFAULT NULL,
  'creatorTypeID' smallint(5) unsigned NOT NULL,
  'orderIndex' smallint(5) unsigned NOT NULL,
  PRIMARY KEY ('itemID','creatorID','orderIndex'),
  KEY 'creatorID' ('creatorID'),
  KEY 'creatorTypeID' ('creatorTypeID'),
  KEY 'name' ('lastName'(7),'firstName'(6))
) ENGINE=InnoDB DEFAULT CHARSET=utf8;

With a many-to-one relationship to items table

File: misc/db-updates/2023-07-17/classic_data_objects.sql (hunk 1 above)
Comment (abaevbog) [reply misc/db-updates/2023-07-17/classic_data_objects.sql:14]: Right. I only kept it as .sql until now for convenience.
I'll have it wrapped into the same algorithm as [this update](https://github.com/zotero/dataserver/blob/master/misc/db-updates/2022-11-05/convertItemIDToBigInt), except with state=`down`.

**Comment by abaevbog** [comment]
In my last commit, I got rid of `creators` table, moved all data storage about creators into the `itemCreators` table, and removed some code dealing with creators that we no longer need. For example, we no longer need the for-loop when we add creators, so I grouped creator insertion into one query. I think there are a few other things can simplify now but that's a start.

**Comment by abaevbog** [comment]
Added a php script to move data from `creators` and `itemCreators` into the new `itemCreators` table on a shard-by-shard basis.

To make it possible to migrate a few shards at a time, I added the php models without changes from this PR with an `old_` prefix. Then, in `ApiController.php`, a global variable `$shardID` is set - based on it and a hardcoded array of shardIDs, `zotero_autoload` in `header.php` decides to use `old_*` files for shards that were not updated yet or usual files for updated shards. I had no problems running tests or using the API after migrating one shard out of 2 locally.

**Comment by abaevbog** [comment]
Brief summary of changes up to this point:

- Added migration script to remove `creators` and `tags` tables and migrate to schema where all data is stored in `itemCreators` and `itemTags`.
- In `header.php`, a temporary global variable `shards` is used to specify which shards have been migrated. Based on this, there is some conditional behavior in `ItemController.php` and `TagController.php`, depending on if the current shard was migrated or not.
- When a method from any affected model is called, `auto_loader` checks current shard against migrated shards and selects either `model/...inc.php` - model with recent changes or `model/old_...inc.php` - model using old schema.
- Both for creators and tags, `load()`, `init()`, and `save()` functions are removed. All data loading happens in the `item` model via `loadCreators` and `loadTags` directly into the object via constructor, and data saving/updating is done via static `bulkCreate` and `bulkDelete`.
- Other minor changes, mainly in table joins, to work with new schema

File: misc/db-updates/2023-07-17/creatorsAsNonClassicDataObjects [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	sleep(60);
+	
+	// Creators 
+	echo "Migrating creators\n";
+	// Drop foreign key constraint
+	Zotero_Admin_DB::query("ALTER TABLE `itemCreators` DROP CONSTRAINT `itemCreators_ibfk_1`;", false, $shardID);
+	Zotero_Admin_DB::query("ALTER TABLE `itemCreators` DROP CONSTRAINT `itemCreators_ibfk_2`;", false, $shardID);
+
+	// Create new itemCreators table
+	Zotero_Admin_DB::query("CREATE TABLE `itemCreatorsNew` ( `creatorID` BIGINT UNSIGNED NOT NULL, `itemID` BIGINT UNSIGNED NOT NULL, `firstName` varchar(255) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci DEFAULT NULL, `lastName` varchar(255) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci DEFAULT NULL, `fieldMode` tinyint(1) UNSIGNED DEFAULT NULL, `creatorTypeID` smallint(5) UNSIGNED NOT NULL, `orderIndex` smallint(5) UNSIGNED NOT NULL, PRIMARY KEY (`creatorID`, `itemID`), KEY `creatorTypeID` (`creatorTypeID`), KEY `name` (`lastName`(7),`firstName`(6)) ) ENGINE=InnoDB DEFAULT CHARSET=utf8;", false, $shardID);
```
Comment (dstillman) [inline misc/db-updates/2023-07-17/creatorsAsNonClassicDataObjects:29]: Sorry, I led you astray originally, and didn't clarify later. I originally said there'd still be `creatorID` in `creators`, but since we're inlining the data in `itemCreators`, there's no reason for there to be a `creatorID` at all — it's not serving any purpose. And the primary key should be on `itemID, orderIndex`, since that's what's unique. (It was actually wrong to begin with — it should've been on `itemID, orderIndex` before like in the client, not `itemID, creatorID, orderIndex`.)

**Review by dstillman — COMMENTED** [review]
Some initial notes

File: misc/db-updates/2023-07-17/creatorsAsNonClassicDataObjects [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+	Zotero_Admin_DB::query("RENAME TABLE itemCreatorsNew TO itemCreators;", false, $shardID);
+
+	// Tags
+	echo "Migrating tags\n";
+	// Drop foreign key constraint
+	Zotero_Admin_DB::query("ALTER TABLE `itemTags` DROP CONSTRAINT `itemTags_ibfk_1`;", false, $shardID);
+	Zotero_Admin_DB::query("ALTER TABLE `itemTags` DROP CONSTRAINT `itemTags_ibfk_2`;", false, $shardID);
+
+	// Create new itemTags table
+	Zotero_Admin_DB::query("CREATE TABLE `itemTagsNew` ( `tagID` BIGINT UNSIGNED NOT NULL, `itemID` BIGINT UNSIGNED NOT NULL, `name` varchar(255) CHARACTER SET utf8mb4 COLLATE utf8mb4_bin NOT NULL, `type` tinyint(1) unsigned NOT NULL DEFAULT '0', `version` int(10) unsigned NOT NULL DEFAULT '1', PRIMARY KEY (`tagID`, `itemID`) ) ENGINE=InnoDB DEFAULT CHARSET=utf8;", false, $shardID);
```
[the remark this answers is not in the document]
Comment (abaevbog) [reply misc/db-updates/2023-07-17/creatorsAsNonClassicDataObjects:51]: 1. Yes, there is a [test](https://github.com/zotero/dataserver/blob/eb3fd636ec0f50eb0bc2ecc310ca1db3e77ed9b1/tests/remote/tests/API/3/TagTest.php#L687) that calls `/tags?newer=`, so I assumed we need to keep it around. If we don't want it - that's great, I'll remove it from here and from my mocha tests PR.

[the remark this answers is not in the document]
Comment (abaevbog) [reply misc/db-updates/2023-07-17/creatorsAsNonClassicDataObjects:51]: 2. Yeah, some context from real world data would certainly be helpful. Maybe if we could see the output of `select count(*) from itemTags` and divide it by `select count(*) from tags`, we'd get an idea of roughly many `itemTag` relationships there are for each tag. And then if we multiply that by `select SUM(length(name)) from tags`, we'd have a rough estimate of how much extra memory duplicating tag names would cause.

[the remark this answers is not in the document]
Comment (abaevbog) [reply misc/db-updates/2023-07-17/creatorsAsNonClassicDataObjects:51]: 3. Queries related to tags that were affected are:

- In `/items?tag=` where we select the itemIDs that fit the tag filter `SELECT itemID FROM items JOIN itemTags USING (itemID) WHERE LOWER(itemTags.name) IN (?)`

- In `tags/?tag=` where we select tagIDs that fit the query before fetching their actual data and constructing objects  `SELECT SQL_CALC_FOUND_ROWS DISTINCT tagID FROM itemTags JOIN items USING (itemID) WHERE libraryID=? AND `name` in (?) AND `type` in (?)`;

Currently, these do a full table scan, which is not good. There was a `UNIQUE KEY uniqueTags (libraryID, name, type)` in the old creators table that was used for such queries. We can add `KEY nameType (name, type)` to current `itemTags` and these queries will use that key. I checked manually with `explain` sql statement and it seems to work.

With creators, the `name` key indeed does not do anything now because the creators never fetched by themselves and are only loaded as a part of an item via `SELECT * FROM itemCreators WHERE itemID=?`. 

To sum up, I guess we should add `KEY nameType (name, type)` to itemTags and remove `name` from creators?
